samples/server/petstore/go-chi-server/go/routers.go

```diff
@@ -159,79 +159,85 @@ func parseTimes(param string) ([]time.Time, error) {
 		if err != nil {
 			return nil, err
 		}
-		times = append(times, t)
+		times = append(times, *t)
 	}
 	return times, nil
 }
 
 // parseTime will parses a string parameter into a time.Time using the RFC3339 format
-func parseTime(param string) (time.Time, error) {
+func parseTime(param string) (*time.Time, error) {
 	if param == "" {
-		return time.Time{}, nil
+		return nil, nil
 	}
-	return time.Parse(time.RFC3339, param)
+	v, err := time.Parse(time.RFC3339, param)
+	return &v, err
 }
 
 type Number interface {
 	~int32 | ~int64 | ~float32 | ~float64
 }
 
-type ParseString[T Number | string | bool] func(v string) (T, error)
+type ParseString[T Number | string | bool] func(v string) (*T, error)
 
 // parseFloat64 parses a string parameter to an float64.
-func parseFloat64(param string) (float64, error) {
+func parseFloat64(param string) (*float64, error) {
```

Comment:
It isn't a good idea to return pointers for all the parsed methods. This more likely than not results in heap allocation and slows it down considerably.

Reply:
A heap allocation will be required at some point for all optional query parameters. My thought process was to keep the mustache template as simple as I could keep it. I tried out not returning a pointer from here, but the changes in the `controller-api.mustache` became quite large.

I will argue that the heap allocation is negligible, especially for the primitive types being handled by these functions. I included two Benchmarks and their results from my local machine below. The summary is that the difference is a 1/1000th of a nanosecond.

I believe that this time penalty is not noticeable and should not be taken into account for how maintainable the code is.


```
// goos: linux
// goarch: amd64
// cpu: Intel(R) Core(TM) i9-9900 CPU @ 3.10GHz
// BenchmarkPointerReturn
// BenchmarkPointerReturn-16    	1000000000	         0.2164 ns/op
// PASS
func BenchmarkPointerReturn(b *testing.B) {
	for i := 0; i < b.N; i++ {
		p := pointerReturn()
		_ = p
	}
}

// goos: linux
// goarch: amd64
// cpu: Intel(R) Core(TM) i9-9900 CPU @ 3.10GHz
// BenchmarkValueReturn
// BenchmarkValueReturn-16    	1000000000	         0.2156 ns/op
// PASS
func BenchmarkValueReturn(b *testing.B) {
	for i := 0; i < b.N; i++ {
		v := valueReturn()
		_ = v
	}
}

func pointerReturn() *int {
	a := 1
	return &a
}

func valueReturn() int {
	a := 1
	return a
}```

Reply:
You need to print allocations.

Also, you need to disable inlining. I suggest you run the benchmark on the parsers in the router.go if you want a better comparison.

Reply:
I just ran the benchmarks on the parsers in router.go and got similar results. The overhead of parsing the strings via `strconv` greatly outweighs any overhead an allocation on the heap that may occur for these primitive types.

```// Pointer return
//goos: linux
//goarch: amd64
//pkg: github.com/SpectraLogic/.../api/openapi
//cpu: Intel(R) Core(TM) i9-9900 CPU @ 3.10GHz
//BenchmarkPointerReturn
//BenchmarkPointerReturn-16    	1000000000	         0.9425 ns/op
//PASS

func BenchmarkPointerReturnBool(b *testing.B) {
	for i := 0; i < b.N; i++ {
		p, _ := parseBool("true")
		_ = p
	}
}

// goos: linux
// goarch: amd64
// pkg: github.com/SpectraLogic/.../api/openapi
// cpu: Intel(R) Core(TM) i9-9900 CPU @ 3.10GHz
// BenchmarkValReturn
// BenchmarkValReturn-16    	1000000000	         0.9367 ns/op
// PASS
func BenchmarkValReturnBool(b *testing.B) {
	for i := 0; i < b.N; i++ {
		p, _ := parseBoolVal("true")
		_ = p
	}
}

// goos: linux
// goarch: amd64
// pkg: github.com/SpectraLogic/.../api/openapi
// cpu: Intel(R) Core(TM) i9-9900 CPU @ 3.10GHz
// BenchmarkPointerReturnFloat64
// BenchmarkPointerReturnFloat64-16    	70159354	        17.99 ns/op
// PASS
func BenchmarkPointerReturnFloat64(b *testing.B) {
	for i := 0; i < b.N; i++ {
		p, _ := parseFloat64("1.0")
		_ = p
	}
}

// goos: linux
// goarch: amd64
// pkg: github.com/SpectraLogic/.../api/openapi
// cpu: Intel(R) Core(TM) i9-9900 CPU @ 3.10GHz
// BenchmarkValReturnFloat64
// BenchmarkValReturnFloat64-16    	59044398	        17.80 ns/op
// PASS
func BenchmarkValReturnFloat64(b *testing.B) {
	for i := 0; i < b.N; i++ {
		p, _ := parseFloat64Val("1.0")
		_ = p
	}
}
```

Reply:
BenchmarkNumNew-8    	48563181	        25.93 ns/op	       4 B/op	       1 allocs/op
BenchmarkBoolNew-8   	100000000	        11.35 ns/op	       1 B/op	       1 allocs/op

BenchmarkNumOld-8    	71088042	        16.85 ns/op	       0 B/op	       0 allocs/op
BenchmarkBoolOld-8   	243797635	         4.883 ns/op	       0 B/op	       0 allocs/op

1.5+x slower on Num (float32)
2+x slower on Bool

inlining turned off

```go
package new

import (
	"testing"
)

var result float32
var result2 bool

func BenchmarkNumNew(b *testing.B) {
	b.ReportAllocs()
	b.ResetTimer()
	for i := 0; i < b.N; i++ {
		r, err := parseNumericParameter[float32](
			"5",
			WithParse[float32](parseFloat32),
			WithMinimum[float32](0),
		)
		if err != nil {
			b.Fatal(err)
		}

		result = *r
	}
}

func BenchmarkBoolNew(b *testing.B) {
	b.ReportAllocs()
	b.ResetTimer()
	for i := 0; i < b.N; i++ {
		r, err := parseBoolParameter(
			"true",
			WithParse[bool](parseBool),
		)
		if err != nil {
			b.Fatal(err)
		}

		result2 = *r
	}
}
```

Reply:
I do understand that it is slower, but it is on the order of a few nanoseconds slower per parameter. 

I have tried a couple of different ways to write the `controller-api.mustache` file without the parsers returning pointers and all of the ways so far have been significantly less maintainable and uglier. Additionally, I haven't been able to smooth out all the wrinkles in the other attempts to get it to work even on my companies REST API Server.